api/v1alpha1/amaltheasession_children.go

```diff
@@ -0,0 +1,220 @@
+
+	sts := appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            cr.Name,
+			Namespace:       cr.Namespace,
+			OwnerReferences: []metav1.OwnerReference{cr.OwnerReference()},
+		},
+		Spec: appsv1.StatefulSetSpec{
+			Replicas: &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+				Spec: v1.PodSpec{
+					Containers:     containers,
+					InitContainers: cr.Spec.ExtraInitContainers,
+					Volumes:        []v1.Volume{{Name: "session", VolumeSource: v1.VolumeSource{PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{ClaimName: pvc.Name}}}},
+				},
+			},
+		},
+	}
+	return sts
+}
+
+// Service returns a AmaltheaSession Service object
+func (cr *AmaltheaSession) Service() v1.Service {
+	labels := labelsForAmaltheaSession(cr.Name)
+
+	svc := v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            cr.Name,
+			Namespace:       cr.Namespace,
+			OwnerReferences: []metav1.OwnerReference{cr.OwnerReference()},
+		},
+		Spec: v1.ServiceSpec{
+			Selector: labels,
+			Ports: []v1.ServicePort{{
+				Protocol:   v1.ProtocolTCP,
+				Name:       "session-port",
+				Port:       80,
+				TargetPort: intstr.FromInt32(cr.Spec.Session.Port),
+			}},
+		},
+	}
+	return svc
+}
+
+// Ingress returns a AmaltheaSession Ingress object
+func (cr *AmaltheaSession) Ingress() *networkingv1.Ingress {
+	labels := labelsForAmaltheaSession(cr.Name)
+
+	ingress := cr.Spec.Ingress
+	if reflect.DeepEqual(ingress, Ingress{}) {
+		return nil
+	}
+
+	path := "/"
+	if ingress.PathPrefix != "" {
+		path = ingress.PathPrefix
+	}
+
+	ing := &networkingv1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            cr.Name,
+			Namespace:       cr.Namespace,
+			Labels:          labels,
+			Annotations:     ingress.Annotations,
+			OwnerReferences: []metav1.OwnerReference{cr.OwnerReference()},
+		},
+		Spec: networkingv1.IngressSpec{
+			IngressClassName: &ingress.IngressClassName,
+			Rules: []networkingv1.IngressRule{{
+				Host: ingress.Host,
+				IngressRuleValue: networkingv1.IngressRuleValue{
+					HTTP: &networkingv1.HTTPIngressRuleValue{
+						Paths: []networkingv1.HTTPIngressPath{{
+							Path: path,
+							PathType: func() *networkingv1.PathType {
+								pt := networkingv1.PathTypePrefix
+								return &pt
+							}(),
+							Backend: networkingv1.IngressBackend{
+								Service: &networkingv1.IngressServiceBackend{
+									Name: cr.Name,
+									Port: networkingv1.ServiceBackendPort{
+										Name: "session-port",
+									},
+								},
+							},
+						}},
+					},
+				},
+			}},
+			TLS: []networkingv1.IngressTLS{{
+				Hosts:      []string{ingress.Host},
+				SecretName: ingress.TLSSecretName,
+			}},
+		},
+	}
+
+	return ing
+}
+
+// PVC returned the desired specification for a persistent volume claim
+func (cr *AmaltheaSession) PVC() v1.PersistentVolumeClaim {
```

Comment:
Isn't the Storage optional ?

Reply:
I will do this in a followup PR. It should be and as you mentioned above we should allow people to use emptyDir or ephemeral volumes or something similar.